lib/gameState/CGameState.h

```diff
@@ -39,12 +39,67 @@ struct SThievesGuildInfo;
 class CRandomGenerator;
 class GameSettings;
 
-struct UpgradeInfo
+class UpgradeInfo
 {
+public:
+	UpgradeInfo() = delete;
+	UpgradeInfo(CreatureID base)
+		: oldID(base), isAvailable(true)
+	{ }
+
 	CreatureID oldID; //creature to be upgraded
-	std::vector<CreatureID> newID; //possible upgrades
-	std::vector<ResourceSet> cost; // cost[upgrade_serial] -> set of pairs<resource_ID,resource_amount>; cost is for single unit (not entire stack)
-	UpgradeInfo(){oldID = CreatureID::NONE;};
+
+	const auto & getAvailableUpgrades() const
+	{
+		return upgradesIDs;
+	}
+
+	const CreatureID & getNextUpgrade() const
+	{
+		return upgradesIDs.back();
+	}
+
+	const auto & getUpgradeCostsFor(CreatureID id) const
+	{
+		auto idIt = std::find(upgradesIDs.begin(), upgradesIDs.end(), id);
+
+		assert(idIt != upgradesIDs.end());
+
+		return upgradesCosts[std::distance(upgradesIDs.begin(), idIt)];
+	}
+
+	const auto & getUpgradeCosts() const
+	{
+		return upgradesCosts;
+	}
+
+	const auto & getNextUpgradeCosts() const
+	{
+		return upgradesCosts.back();
+	}
```

Comment:
What exactly this does? What is "next upgrade" and why it always picks last element from list?

Reply:
It used to pick first element in code previous to [this PR](https://github.com/vcmi/vcmi/pull/4348/commits/6cf423d97760553ff7d32f3cea814598ce8d6331), but works correctly when it takes the last one up untill there are no upgrades. UpgradeInfo is refilled on every upgrade attempt. 
I tested this code to work with alternative upgrades and with third-upgrades-mod that allows to do upgrades two times from basic. This `getNextUpgradeCosts` will pick the upgrade costs of upgrade which should be done at that specific moment. 
Hmmm... I have to admit: "Next" is rather misleading and the naming should be just `getUpgradeCosts` and maybe the other function with that name should be `getAvailableUpgradesCosts`.

Reply:
Yeah, `getUpgradeCosts` sounds more clear